src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs

```diff
@@ -10,4 +10,513 @@
+onOneArgs op f typ instrs = do
+  v ← traverse (protect . (inst >=> promoteTopStack)) instrs
+  case v of
+    instr1 : _ → do
+      res ←
+        if
+          | allConstants [val instr1] →
+            let Env.Constant i1 = val instr1
+             in pure (f i1)
+          | otherwise → do
+            addExpanded instr1
+            addInstr op
+            pure (Env.Expanded op)
+      modify @"stack" (VStack.drop 2)
+      consVal res typ
+      pure res
+    _ → throw @"compilationError" Types.NotEnoughArguments
+
+-------------------------------------------------------------------------------
+-- Environment Protections, Promotions, and Movements
+--------------------------------------------------------------------------------
+
+moveToFront ∷ (Env.Instruction m, Integral a) ⇒ a → m Instr.ExpandedOp
+moveToFront num = do
+  let inst = Instructions.dig (fromIntegral num)
+  addInstr inst
+  modify @"stack" (VStack.dig (fromIntegral num))
+  pure inst
+
+dupToFront ∷ (Env.Instruction m, Integral a) ⇒ a → m Instr.ExpandedOp
+dupToFront num = do
+  modify @"stack" (VStack.dupDig (fromIntegral num))
+  let instrs =
+        [ Instructions.dig (fromIntegral num),
+          Instructions.dup,
+          Instructions.dug (fromIntegral num)
+        ]
+  addInstrs instrs
+  pure (fold instrs)
+
+data Protect
+  = Protect
+      { val ∷ Env.Expanded,
+        insts ∷ [Types.Op]
+      }
+
+protect ∷ Env.Ops m ⇒ m Env.Expanded → m Protect
+protect inst = do
+  curr ← get @"ops"
+  v ← inst
+  after ← get @"ops"
+  put @"ops" curr
+  pure Protect {val = v, insts = after}
+
+-- Promoting types happen elsewhere
+-- so protect just serves to hold the ops effects
+addExpanded ∷ Env.Ops m ⇒ Protect → m ()
+addExpanded (Protect _ i) = addInstrs i
+
+promoteTopStack ∷ Env.Instruction m ⇒ Env.Expanded → m Env.Expanded
+promoteTopStack x = do
+  stack ← get @"stack"
+  let (insts, stack') = VStack.promote 1 stack
+  put @"stack" stack'
+  addInstrs insts
+  pure x
+
+reserveNames ∷ HasState "count" Word m ⇒ Word → m [Symbol]
+reserveNames i = do
+  c ← get @"count"
+  put @"count" (i + c)
+  pure (intern . show <$> [c .. c + i - 1])
+
+-- TODO ∷ drop extra things from the vstack, mainly
+-- 1. the function we move to the front
+-- 2. the expanded lambda we create
+-- Note that it isn't vital, as these aren't stored in the real Michelson stack
+-- Other things considered:
+-- We don't need to drop the arguments we eval and name, as they should be eaten
+-- by the functions they call with the appropriate usages
+apply ∷ Env.Reduction m ⇒ Env.Curr → [Types.NewTerm] → [Symbol] → m Env.Expanded
+apply closure args remainingArgs = do
+  let totalLength = fromIntegral (length args + length remainingArgs)
+  case totalLength `compare` Env.left closure of
+    EQ → do
+      -- usage and type doesn't matter here!
+      evalArgsAndName
+      app
+    LT → do
+      evalArgsAndName
+      -- make new closure, and apply args to the fun in the closure
```

Comment:
Odd phrasing, you mean evaluate the arguments which have been applied so far and then return a closure which will take the remaining arguments then inline the function, correct?

Reply:
Well maybe inline, it depends if we have enough arguments by the top level whether we can inline or not, but yes correct.